### ISSUE TITLE
feedback for softwareFeature.schema.tpl.json

### ISSUE THREAD

**Issue by lzehl** [openMetadataInitiative/openMINDS_core#142]
@bweyers: please have a look at the [softwareFeature.schema.tpl.json](https://raw.githubusercontent.com/HumanBrainProject/openMINDS_core/v3/schemas/products/softwareFeature.schema.tpl.json)

My questions:
1) Does it make sense to have this schema in **products** or should it move into an other folder?
2) What do you link to in the property **definitionReference** (previously "definition")? Is this a link to an ontology? The reason why we changed the property name is that the property "definition" in all other cases it is used expects a string (with a one sentence definition). I don't like the term "definitionReference", but I was not sure what is expected here (behind the IRI). Could you explain?

**Comment by jagru20** [comment]
Hey, I hope it's okay if I jump in here.
Referring **point 1:** Placing the softwareFeature.schema.tpl.json under "products" discussable, as it is not a research product as such. It has similarities with the controlledTerm schema, with the difference that the ontology used in the controlledTerm schema are maintained by the openMINDS team whereas the definition link in the softwareFeature  schema refers to an external ontology. For example the link for the feature "platform independent" could be https://www.wikidata.org/wiki/Q174666. 
This leads to **point 2:** Renaming the property **definition** property into **definitionReference** makes this clearer and is ok for us.

Long story short: 
The sofwareFeature schema could be better placed under "miscellaneous" or "controlledTerm" (which one do you prefer?) and renaming the property **definition** property into **definitionReference** should be fine.

**Comment by lzehl** [comment]
@jagru20 yes of course your feedback is appreciated!

Okay so if I understand correctly the "definitionReference" links always to an ontology?

If that is the case I would indeed suggest to define this schema as a controlledTerm.

The controlledTerms are organized by terminology / category (e.g. BiologicalSex, Species, ModelScope, etc) I guess this would then be the software feature categories that need to be used here to group the software features?

Each term has the properties:
- name
- definition (string, one sentence) 
- description (string, longer description) 
- ontologyIdentifier (IRI to ontology term)

Would that fit?

**Comment by lzehl** [comment]
@jagru20 & @bweyers I urgently need feedback for the required changes for software feature for the release (which has to happen today...)

If you do not object I will do the following changes:
- SoftwareFeature schema will be deleted from core
- in SoftwareVersion schema property "hasFeature" will link to controlledTerms for softwareFeatures 

For the latter I have an additional question:
From what I see the only software feature category currently in use is "abstract". 
Are the others in the KG deprecated (and there is only going to be this one category) or are the others only not used yet?
1) If the first is the case I would just create a controlledTerms schema called "SoftwareFeature" and list all KG registered features in the corresponding terminology. 
2) If the second is the case I would create a controlledTerms schema called "AbstractSoftwareFeature" and list all KG registered features in there that have the category "abstract". If the other feature categories are ever be used they can be added as new controlledTerms schemas and corresponding terminologies (e.g., "DisplaySoftwareFeature").

Please let me know asap what option you would prefer (I went for option 2 right now).

**Comment by bweyers** [comment]
@lzehl Hi!
In general, handling SoftwareFeature as controlled term is OK! Makes a lot of sense as we are do it similarly in the current moment. As far as I rember, it was also possible to define hierarchical strcutures there? Our current list of features is here: https://github.com/bweyers/HBPVisCatalogue/wiki/Identified-Software-Features 

For your question due to the use of SWFeatures in the current KG structure, this seams odd... Actually, we only have features we also actually use. Please see the link above.

**Comment by lzehl** [comment]
@bweyers thanks for this overview. It makes more sense to me now. Apparently the only software feature category that is used in the KG at the moment is Abstract Data Types (but in KG only listed as "abstract"). 

Which of the following options would you prefer (considering this category grouping)?
(Note that if the hierarchy of software features is defined in the linked ontology, I would prefer option A)

OPTION A)
- controlledTerms/schemas/  
  - softwareFeature.schema.tpl.json
- controlledTerms/terminologies/softwareFeature/
  - GraphDataTypes.jsonld
  - StatisticalDataTypes.jsonld
  etc.

OPTION B)
- controlledTerms/schemas/  
  - softwareFeatureDataType.schema.tpl.json
  etc.
- controlledTerms/terminologies/softwareFeatureDataType/
  - GraphDataTypes.jsonld
  - StatisticalDataTypes.jsonld
  etc.

OPTION C)
- controlledTerms/schemas/  
  - softwareFeatureAbstractDataType.schema.tpl.json
  etc.
- controlledTerms/terminologies/softwareFeatureAbstractDataType/
  - GraphDataTypes.jsonld
  - StatisticalDataTypes.jsonld
  etc.

**Comment by bweyers** [comment]
@lzehl Thus, what you mean is that we are defining the hierarchy in softwareFeature.schema.tpl.json?

**Comment by bweyers** [comment]
@lzehl Just discussing with Jan: Thus, we would represent our hierarchy as follows. Would this makes sense? The hierarchy is not represented in WikiData, which we normally use as defining reference of the lowest hierarchy. Of course, the structure should be consistent to other structures. I do not want to generate a new thing here only for SoftwareFeatures.

- controlledTerms/terminologies/softwareFeature/AbstractDataTypes/
   -  GraphDataTypes.jsonld
   -  StatisticalDataTypes.jsonld
   -  ...

- controlledTerms/terminologies/softwareFeature/2DImageDataTypes/
  -  RasterImageDataTypes.jsonld
  -  VectorImageDataTypes.jsonld

- controlledTerms/terminologies/softwareFeature/3DScientificDataTypes/
  -  3DScalarDataTypes.jsonld
  -  ...

- controlledTerms/terminologies/softwareFeature/UsageTypes/

- controlledTerms/terminologies/softwareFeature/DisplayTypes/

- controlledTerms/terminologies/softwareFeature/UserInterfaceTypes/

- controlledTerms/terminologies/softwareFeature/SystemArchitectureTypes/

- controlledTerms/terminologies/softwareFeature/DevelopmentTypes/

**Comment by lzehl** [comment]
@bweyers no what you suggest is not possible within controlledTerms. The options I gave you above is what can be done within controlledTerms. The grouping was only introduced to be able to reference specific sets of terms within one property (e.g., property "ageCategory" allows all controlledTerms of type "AgeCategory").  Everything else (such as relations and hierarchies) is for other terms handled in the correspondingly attached ontologies.

If you prefer subcategories for the SoftwareFeatures within the controlledTerms, I would need to implement option C and then list all possible controlledTerms types for software features in the property "hasFeature" in the SoftwareVersion schema.

**Comment by bweyers** [comment]
@lzehl ok, thanks for clarifying. We would then go for option A) and using WikiData as main source for hierarchies and terms necessary. This would be the simplest and easiest way. If we start to include things into the SoftwareVersion schema we would gain a certain level of infexibility, which should be avoided!

**Comment by lzehl** [comment]
It seems we clarified the open questions in this issue. I'm closing it for now. s
